src/example_behaviors/src/sam2_segmentation.cpp

```diff
@@ -0,0 +1,137 @@
+
+namespace
+{
+  constexpr auto kPortImage = "image";
+  constexpr auto kPortImageDefault = "{image}";
+  constexpr auto kPortPoint = "pixel_coords";
+  constexpr auto kPortPointDefault = "{pixel_coords}";
+  constexpr auto kPortMasks = "masks2d";
+  constexpr auto kPortMasksDefault = "{masks2d}";
+
+  constexpr auto kImageInferenceWidth = 1024;
+  constexpr auto kImageInferenceHeight = 1024;
+} // namespace
+
+namespace example_behaviors
+{
+  SAM2Segmentation::SAM2Segmentation(const std::string& name, const BT::NodeConfiguration& config,
+                                     const std::shared_ptr<moveit_studio::behaviors::BehaviorContext>& shared_resources)
+    : moveit_studio::behaviors::AsyncBehaviorBase(name, config, shared_resources)
+  {
+
+    std::filesystem::path package_path = ament_index_cpp::get_package_share_directory("example_behaviors");
+    const std::filesystem::path encoder_onnx_file = package_path / "models" / "sam2_hiera_large_encoder.onnx";
+    const std::filesystem::path decoder_onnx_file = package_path / "models" / "decoder.onnx";
+    sam2_ = std::make_shared<moveit_pro_ml::SAM2>(encoder_onnx_file, decoder_onnx_file);
+  }
+
+  BT::PortsList SAM2Segmentation::providedPorts()
+  {
+    return {
+      BT::InputPort<sensor_msgs::msg::Image>(kPortImage, kPortImageDefault,
+                                             "The Image to run segmentation on."),
+      BT::InputPort<std::vector<geometry_msgs::msg::PointStamped>>(kPortPoint, kPortPointDefault,
+                                                                   "The input points, as a vector of <code>geometry_msgs/PointStamped</code> messages to be used for segmentation."),
+
+  BT::OutputPort<std::vector<moveit_studio_vision_msgs::msg::Mask2D>>(kPortMasks, kPortMasksDefault,
+                                                                   "The masks contained in a vector of <code>moveit_studio_vision_msgs::msg::Mask2D</code> messages.")
+
+
+    };
+  }
+
+  void SAM2Segmentation::set_onnx_image_from_ros_image(const sensor_msgs::msg::Image& image_msg,
+                                                       moveit_pro_ml::ONNXImage& onnx_image)
+  {
+    onnx_image.shape = {1, image_msg.height, image_msg.width, 3};
+    onnx_image.data.resize(image_msg.height * image_msg.width * 3);
+     int stride = image_msg.encoding != "rgb8" ? 3: 4;
+    for (size_t i = 0; i < onnx_image.data.size(); i+=stride)
+    {
+      onnx_image.data[i] = static_cast<float>(image_msg.data[i]) / 255.0f;
+      onnx_image.data[i+1] = static_cast<float>(image_msg.data[i+1]) / 255.0f;
+      onnx_image.data[i+2] = static_cast<float>(image_msg.data[i+2]) / 255.0f;
+    }
+  }
+
+
+  void SAM2Segmentation::set_ros_image_from_onnx_image(const moveit_pro_ml::ONNXImage& onnx_image, sensor_msgs::msg::Image& image_msg)
+  {
+    image_msg.height = static_cast<uint32_t>(onnx_image.shape[0]);
+    image_msg.width = static_cast<uint32_t>(onnx_image.shape[1]);
+    image_msg.encoding = "mono8";
```

Comment:
So, is the onnx image in this function assumed to be greyscale, but in the other one rgb? Seems like  at least the function names (which implied to me there inverses of each other) need different names.

Reply:
I renamed the function to `set_ros_mask_from_onnx_mask`. Now it is more clear that it converts a 1 channel onnx image to a mask.